content/concepts/hole-punching.md

```diff
@@ -0,0 +1,193 @@
+it may not be available everywhere, and there can posses the risk
+of establishing connections with untrustworthy nodes as the protocol
+does not enforce authentication or authorization. Manually opening a
+port requires technical expertise and does not enforce 
+authentication or authorization. Using a relay node also requires
+technical expertise. Relaying adds additional latency and is resource
+intensive.
+
+### Possible solution: hole punching
+
+In the case where the other options aren't sufficient, networks can 
+use a technique called hole punching to establish connections with 
+non-public nodes.
+
+Each node connects to an unrestricted third-party server and 
+shares its external and internal address and port information.
+The server temporarily stores the node's information and relays 
+each node's information to the other. Clients can use this 
+information to establish direct connections with each other.
+
+Take two nodes, `A` and `B`, that would like the dial each other:
+
+1. The first packet of both nodes (e.g., in the case of TCP, an SYN) 
+   passes through their respective routers.
+2. The routers add a 5-tuple to their router's state table. 
+3. `Packet A` and `B` "punch holes" into their respective routers' 
+   firewalls.
+4. Both packets arrive at the opposite router.
+5. Once `A`'s packet arrives at `router B`, `router B` checks its state 
+   table and finds a 5-tuple previously added through the packet sent by 
+   node B. 
+6. The routers forward the packets through the "punched holes". 6. It 
+   then forwards the packet to `B`. The same occurs with `B`'s packet; 
+   upon arriving at `router A`, it matches a 5-tuple in `router A`'s state 
+   table and thus forwards the packet to `A`.
+  
+The following use case diagram illustrates the above process.
+
+![](https://i.imgur.com/0k2Zlj3.png)
+
+{{% notice "note" %}}
+This process assumes a mechanism to synchronize `A` and `B` simultaneously.
+{{% /notice %}}
+
+## Hole punching in libp2p
+
+Inspired by the 
+[ICE protocol](https://datatracker.ietf.org/doc/html/rfc8445) 
+specified by the IETF, libp2p includes a decentralized hole punching 
+feature that allows for firewall and NAT traversal without the need 
+for central coordination servers like STUN and TURN. 
+
+The following sequence diagram illustrates the whole process.
+
+![](https://i.imgur.com/sMGMfGZ.png)
+
+libp2p hole punching can be derived in two phases, a preparation phase and 
+a hole punching phase.
+
+### Phase I: Preparation
+
+1. [AutoNAT](/concepts/nat/#autonat): Determine whether a node is dialable, 
+   as in, discover if a node is behind a NAT or firewall.
+   
+   > This is equivalent to the 
+   > [STUN protocol](https://www.rfc-editor.org/rfc/rfc3489) in ICE.
+
+   ![](https://i.imgur.com/NeufUm7.png)
+   
+   - `B` reaches out to `Other_Peers` (e.g., boot nodes) on the network it 
+     is on and asks each node to dial it on a set of addresses it suspects 
```

Comment:
This should probably mention that a libp2p node has multiple ways of discovering its addresses, but the most important one is Identify (link to concept doc), our STUN equivalent.